### ISSUE TITLE
Select test cases for a new session

### ISSUE THREAD

**Issue by 32leaves** [32leaves/ruruku#76]
One does not always want to test all testcases in a single testsuite. Sometimes one only wants to run a very basic test or run all tests for a specific feature. Ruruku should support something like a `testset` which is selection of testcases from a testsuite. During the creation of a session one should be able to select a testset (or a subset of testcases directly).

For this feature we'd need to introduce three new concepts:
1. *testcase annotations*: annotations are key/value pairs attached to a testcase (e.g. `feature=signup`, `concurrentUsers=5` or `dependsOnBrowser=true`. The values of annotations are either symbols (strings without spaces) or numeric values.
2. *selection expressions*: based on the annotations we want to select testcases. An expression evaluates to a boolean value for each testcase depending on if the expression matches the testcase's annotations or not. For example: `feature == signup and (concurentUsers > 2 or dependsOnBrowser == true)` The expression language is as follows:
```
SYMBOL = /\w[\w\d]+/
NUMBER = /\d+(\.\d+)?/
OPERATOR = '==' | '!=' | '>' | '>=' | '<' | '<='
VALUE = SYMBOL | NUMBER
SINGLE_EXPR = SYMBOL OPERATOR VALUE
CONJUNCTION = 'and' | 'or'
MULTI_EXPR = SINGLE_EXPR (CONJUNCTION SINGLE_EXPR)*
EXPR = PARANTHESIZED_EXPR | MULTI_EXPR
PARANTHESIZED_EXPR = '(' EXPR ')'
```
3. In a testsuite one can have testsets which have an `ID`, a `name` and an `expression`. During the creation of a test session one can select a testset or explicitly give an expression. The test session then only contains all tests that match the expression. Additionally, the testsession will have an annotation which contains the testset name and the expression.

This issue is best split in three tasks along the individual features.

- [ ] Implement expression language incl. interpreter
- [ ] Add annotations to testcases and testsets to testsuites
- [ ] Implement selection during session creation

**Comment by 32leaves** [comment]
We don't have to implement the expression evaluation ourselves, but use github.com/Knetic/govaluate instead.

**Comment by bwl21** [comment]
maybe it could be the other way round ...

1. a variant rich test suite `TV` is
    1. `V` a set of variation selectors, where each selector has a name and a value range `VS1 ... VSn`
    2. `T` a set of test cases, `T1 .. Tn` each hast a variation point. A variation point is manifested as a binding expression `boolean fbind = f(S1 ... Sn)`

   `TV = {V, T}`
    
2. a bound test suite `Tb` is then

   `Tb =  f(TV, VD)` with `VD = {VS1: <value1> ... VSn: <valuen>}`

    `Tb = TV.T .select {|t| fbind(VD) == true}`

    In other words the bound Testsuite is a subset of the variant rich testsuite containing all 
    testcases for which the binding expression yields true for a given variant vector.

3. for practical reasons it might make sense to factor out the binding expression from the testcases such that they can be reused.

**Comment by 32leaves** [comment]
I think testsets (testcase selection) and testcase variations are orthogonal: `testset(variant(allTestCases))`.  A testset selects a subset based of testcases. The use-case is that one wants to run a test based on a reduced set of tests. Test variants enable variability in test-selection and values. The use-case is that one wants to run tests for varying environments or configurations. One may however want to run a reduced yet variable testplan.

For test sets I think it makes most sense to implement them on a testplan level (incl. their "binding expressions"). For variations the method you describe is the way to go.
